\documentclass[a4paper,twocolumn]{article}
\usepackage[hidelinks]{hyperref}

\title{Introduction to data analysis for natural and social sciences}
\author{Marco Casari}
\date{\today}

\begin{document}
\maketitle

\section{Introduction}
The present document constitutes the second part of the exam. A summary of article ``Inferring the immune response from repertoire sequencing'' is provided.%
\footnote{Version 2 of the article is referenced. Full citation: Puelma Touzel M, Walczak AM, Mora T (2020) Inferring the immune response from repertoire sequencing. PLoS Comput Biol 16(4): e1007873. \url{https://doi.org/10.1371/journal.pcbi.1007873}.}
Results sections ``Modeling repertoire variation'' and ``Inferring the noise profile from replicate experiments'' are supplied with technical results.

% MC possible roadmap:
%content
%affecting
%methodology
%validity of predictions
%future development

% MC in general: introduction + summary, Results as subsections.
%problem
%how they address the problem
%how they validate their solution

% MC also: (1) what, (2) why, (3) how.
\section{Summary}
% MC I do first the subsections since they are technical and I will insert also figures then I proceed to insert the relevant information in this part.
The article introduces a probabilistic model able to estimate the noise and describe clonotypes expansion in longitudinal Repertoire Sequencing (RepSeq) data.
Next Generation Sequencing (NGS) provides large amounts of RepSeq data, which standard inference methods fails to describe accurately due to experimental noise and biological diversity. Many factors contribute to the noise: some are related to the experiment execution (e.g. sampling procedure, library preparation), others have biological origin (e.g. gene expression). These factors introduce variability in sequence counts which is not related to the natural variability of T-Cell Receptors (TCRs).

The model presented in the article is able to decouple the noise distribution from the clonotype counts distribution, learning parameters of both from RepSeq data. Additionally, once the parameters are learned, a Bayesian approach can be used to ...
% MC moreover the model can then be used for inference.
% The model can be used to predict the dynamics of individual clonotypes using Bayesian inference to get the posterior probability of expansion.

% MC how they validate their inference experiment using the model. vaccination against yellow fever.

% MC to which data the model can be applied? mRNA data, but also to gDNA with slight variations, see results.

\subsection{Modeling repertoire variation}
The model consists in three main parts, identified by: noise model, clone size distribution and dynamical model.

The noise model for an individual clonotype is a conditional probability $P(n|f)$ of cell counts $n$ given the true frequency $f$. This definition is necessary because frequency $f$ is not known and RepSeq experiments supply a TCR read count (i.e. a proxy for cell count) which depends on the frequency.
% MC remove this part if useless.
%Index $i = 1, \dots, N$ is a label to identify each clonotype, with $N$ total number of clonotypes in the immune system. This value is not known either, the total number of counts $N_\mathrm{read}$ is used instead.%
%\footnote{Clonotype index is omitted when the meaning of variables is clear.}
Three probability distributions have been considered for the noise model: Poisson distribution, negative binomial and a mixed negative binomial-Poisson model. Poisson distribution is set with mean $\bar{n} = f N_\mathrm{read}$ where $N_\mathrm{read}$ is the total number of counts, but it is not effective in reproducing observed noise because of its smaller variance. The negative binomial distribution has mean $\bar{n}$ and variance $\bar{n} + a \bar{n}^\gamma$, where the two parameters $a$ and $\gamma$ are used to learn the noise variance. The third model considers explicitly the cell count $m$ of the clonotype in the sample used to infer the general model. It is built by summing over $m \in \{1, \dots, M\}$ the product of the negative binomial $P(m|f)$ with mean $f M$ and the Poisson distribution $P(n|m)$ with mean $\frac{m}{M} N_\mathrm{read}$, where $M$ is the total number of cells in the sample, which is an additional parameter of the general model. This distribution is preferred because it has the best performance on RepSeq data.

The clone size distribution $\rho(f)$ of a given clonotype is a power law with exponent $\nu$. To avoid divergence, a minimum frequency $f_\mathrm{min}$ is set and its value corresponds to the frequency of lymphocytes if each one were an individual clonotype. There is one constraint on clonotype frequencies, namely they must sum up to 1, and it is satisfied when number of clonotypes $N$ is large and $\langle f \rangle = \frac{1}{N}$.
The joint distribution of frequencies is the product of $\rho(f)$ over the entire repertoire and it is set non-zero when the constraint holds. Normally the constraint denies immediate factorisation of the joint distribution, except when it is verified, for instance during the sampling procedure when it is enforced.

The dynamical model $G(f', t'|f, t)$ describes the probability of expansion or contraction of a clonotype from frequency $f$ at time $t$ to frequency $f'$ at a successive time $t'$. For the purposes of the article, $t$ is before vaccination and $t'$ is after vaccination. To explicit $G(f', t'|f, t)$, it is assumed that a fraction $\alpha$ of clonotypes reacts to a stimulus and that expansion or contraction of a clonotype does not depend on its size. To quantify the response, a log fold change $s = \ln \big( \frac{f'}{f} \big)$ is defined and the infinitesimal variation in frequency of the dynamical model is expressed as infinitesimal variation of function
\[
  \rho_s(s) = (1 - \alpha) \delta(s - s_0) + \alpha \rho_\mathrm{exp} (s - s_0) \quad ,
\]
where $s_0 < 0$ is a measure of contraction extended to the entire repertoire to compensate for the expansion and preserve the constraint on frequencies, while $\rho_\mathrm{exp}$ modulates the expansion of activated clonotypes. The form of $\rho_\mathrm{exp}$ is chosen according to the case under analysis.

Parameters are evaluated in two consecutive steps. First, replicate RepSeq experiments are performed at time $t$ and parameters of distributions $P(n|f)$ and $\rho(f)$ are inferred simultaneously. Second repertoires at different times $t$ and $t'$ are compared to evaluate the parameters of distribution $G(f', t'|f, t)$.

The possibility to work in a Bayesian framework is one of the differences of the model when compared to standard approaches of differential expression analysis. Others are the intrinsic treatment of finite cell counts given an activated clonotype, instead of working with average gene expression data, and the use % MC non mi piace, trovare altro verbo.
of $\rho(f)$ as prior for the evaluation of the likelihood of expansion.

\subsection{Inferring the noise profile from replicate experiments}

% MC another image to use to discuss their results and to insert in the summary is S1 fig, cfr. page 5.
% MC in realtÃ  l'immagine fa schifo, poco interessante, non descriverei questa.

% MC hill diversities can be estimated.

\end{document}
